### PR TITLE
Translations: force json, convert template, default to existing key

### PR DESCRIPTION
**Pull request by Moulberry**
Adds 3 new translation features (I can split into 3 PRs if needed)

1. Add Force JSON Translation
For me (and maybe others), the plugin isn't able to detect the Minecraft version and throws "Cannot determine MC version". This option makes it avoid erroring

2. Add template for the Convert to Translation feature
net.minecraft.client.resources.I18n isn't always desired, this option allows the template string to be changed

3. Default to existing key when adding duplicate translation text
Scans the translation file for an existing key that matches the translation text and suggests it inside the Convert to Translation dialog

Feature 3 only supports json translation files, support for lang files can be added if desired.

Apologies if there are any stylistic issues, I am not a Kotlin developer

### CODE REVIEW

File: src/main/kotlin/MinecraftSettings.kt [hunk 1]
```diff
@@ -37,6 +37,9 @@ class MinecraftSettings : PersistentStateComponent<MinecraftSettings.State> {
         var underlineType: UnderlineType = UnderlineType.DOTTED,
 
         var isShadowAnnotationsSameLine: Boolean = true,
+
+        var isForceJsonTranslationFile: Boolean = false,
+        var convertToTranslationTemplate: String = "net.minecraft.client.resources.I18n.format(\"\$key\")"
```
Comment (LlamaLad7) [inline src/main/kotlin/MinecraftSettings.kt]: Trailing comma

File: src/main/kotlin/translations/TranslationFiles.kt [hunk 2]
```diff
@@ -223,6 +260,17 @@ object TranslationFiles {
...
     }
 
+    private fun PsiFile.findKeyForTextAsJson(text: String): String? {
+        val rootObject = this.firstChild as? JsonObject ?: return null
+        return rootObject.propertyList.firstOrNull {
+            if (it.value is JsonStringLiteral) {
+                (it.value as JsonStringLiteral).value == text
+            } else {
+                false
+            }
```
Comment (LlamaLad7) [inline src/main/kotlin/translations/TranslationFiles.kt]: Can be `(it.value as? JsonStringLiteral)?.value == text`

File: src/main/kotlin/translations/intentions/ConvertToTranslationIntention.kt [hunk 3]
```diff
@@ -73,12 +77,14 @@ class ConvertToTranslationIntention : PsiElementBaseIntentionAction() {
             val key = result.first ?: return
             val replaceLiteral = result.second
             try {
-                TranslationFiles.add(element, key, value)
+                if (existingKey == null || existingKey != key) {
```
Comment (LlamaLad7) [inline src/main/kotlin/translations/intentions/ConvertToTranslationIntention.kt]: No need for the null check, `==` is null-safe.

File: src/main/kotlin/translations/TranslationFiles.kt [hunk 4]
```diff
@@ -110,12 +111,48 @@ object TranslationFiles {
...
+
+        for (file in files) {
+            val psiFile = PsiManager.getInstance(context.project).findFile(file) ?: continue
+            psiFile.apply {
+                if (jsonVersion) {
+                    this.findKeyForTextAsJson(text)?.let { return it }
+                } else {
+                    // todo: this.findKeyForTextAsLang(text)
+                }
+            }
```
Comment (LlamaLad7) [inline src/main/kotlin/translations/TranslationFiles.kt]: Not sure I'd go for the `apply` over just `psiFile.findKeyForTextAsJson(text)`

**Review by LlamaLad7 — CHANGES_REQUESTED** [review]

File: src/main/kotlin/translations/TranslationFiles.kt (hunk 4 above)
Comment (Moulberry) [reply src/main/kotlin/translations/TranslationFiles.kt]: [a135ee5](https://github.com/minecraft-dev/MinecraftDev/pull/2292/commits/a135ee54cdb3eb9d910636ea014f32906314292d)

File: src/main/kotlin/translations/intentions/ConvertToTranslationIntention.kt (hunk 3 above)
Comment (Moulberry) [reply src/main/kotlin/translations/intentions/ConvertToTranslationIntention.kt]: [1ab56e9](https://github.com/minecraft-dev/MinecraftDev/pull/2292/commits/1ab56e9784fd3512976b4f0963019ada86ac9f3a)

File: src/main/kotlin/translations/TranslationFiles.kt (hunk 2 above)
Comment (Moulberry) [reply src/main/kotlin/translations/TranslationFiles.kt]: [271746a](https://github.com/minecraft-dev/MinecraftDev/pull/2292/commits/271746a3b0fea06c09f033fef527351451d25505)

File: src/main/kotlin/MinecraftSettings.kt (hunk 1 above)
Comment (Moulberry) [reply src/main/kotlin/MinecraftSettings.kt]: [97988d6](https://github.com/minecraft-dev/MinecraftDev/pull/2292/commits/97988d6171f106b430b48a6ee7eebedb0931add5)

File: src/main/kotlin/MinecraftConfigurable.kt [hunk 5]
```diff
@@ -91,6 +92,16 @@ class MinecraftConfigurable : Configurable {
             }
         }
 
+        group(MCDevBundle("minecraft.setting.translation")) {
```
Comment (RedNesto) [inline src/main/kotlin/MinecraftConfigurable.kt]: Nitpick, same for the two other keys below

```suggestion
        group(MCDevBundle("minecraft.settings.translation")) {
```

File: src/main/kotlin/MinecraftSettings.kt [hunk 6]
```diff
@@ -37,6 +37,9 @@ class MinecraftSettings : PersistentStateComponent<MinecraftSettings.State> {
         var underlineType: UnderlineType = UnderlineType.DOTTED,
 
         var isShadowAnnotationsSameLine: Boolean = true,
+
+        var isForceJsonTranslationFile: Boolean = false,
+        var convertToTranslationTemplate: String = "net.minecraft.client.resources.I18n.format(\"\$key\")",
```
Comment (Earthcomputer) [inline src/main/kotlin/MinecraftSettings.kt]: I believe intentions can have their own settings, so no need to make this a global setting. Although in this case we could probably get away with it being hardcoded if you use `MappingsManager`. Right now you would need to add it to `HardcodedYarnToMojmap` as we haven't fully implemented the mappings API yet.

It would be nice to figure out why we're unable to detect the MC version rather than forcing it. But this is probably good enough for a temporary solution.

**Review by Earthcomputer — CHANGES_REQUESTED** [review]

**Comment by Earthcomputer** [comment]
We have already dropped support for new features for MC 1.12, so I don't mind support for lang files being unimplemented.

File: src/main/kotlin/MinecraftSettings.kt (hunk 6 above)
Comment (Moulberry) [reply src/main/kotlin/MinecraftSettings.kt]: Its not just about mapping from yarn to mojmap, I personally like to use a small wrapper that can do some additional formatting. People on other platforms (paper, etc.) might also want to leverage the intention without being forced to use NMS. If you still want it to go through HardcodedYarnToMojmap let me know and I'll make the change

I'll look into how to move the settings into the intention settings

Comment (Earthcomputer) [reply src/main/kotlin/MinecraftSettings.kt]: That's a fair enough reason, although I think the usage of yarn is common enough that we need to be able to make it work without any config changes. I guess just leave the remapping problem for now and we'll fix it once we get the mappings stuff up and running.

Comment (Moulberry) [reply src/main/kotlin/MinecraftSettings.kt]: I'll try see if I can add an explicit checkbox for "Use Custom Template .. etc", which should make it easier to implement that behaviour in the future.

Comment (Moulberry) [reply src/main/kotlin/MinecraftSettings.kt]: It was found that intentions cannot have their own settings. Leaving the options in MinecraftConfigurable seems like the easiest solution right now

File: src/main/kotlin/MinecraftConfigurable.kt [hunk 7]
```diff
@@ -97,8 +99,17 @@ class MinecraftConfigurable : Configurable {
                 checkBox(MCDevBundle("minecraft.settings.translation.force_json_translation_file"))
                     .bindSelected(settings::isForceJsonTranslationFile)
             }
-            row(MCDevBundle("minecraft.settings.translation.convert_to_translation_template")) {
+
+            lateinit var property: ComponentPredicate
```
Comment (RedNesto) [inline src/main/kotlin/MinecraftConfigurable.kt]: This variable deserves a better name.

Also not a fan of having this option in an application-wide configurable, its value will vary based on the project. It would be better to have it in a project-level configurable.

File: src/main/kotlin/MinecraftConfigurable.kt (hunk 5 above)
Comment (Moulberry) [reply src/main/kotlin/MinecraftConfigurable.kt]: [b3bd4e7](https://github.com/minecraft-dev/MinecraftDev/pull/2292/commits/b3bd4e76f7057f9dab92acc49449ad39e3f450a4)

File: src/main/kotlin/MinecraftConfigurable.kt (hunk 7 above)
Comment (Moulberry) [reply src/main/kotlin/MinecraftConfigurable.kt]: [d1050e8](https://github.com/minecraft-dev/MinecraftDev/pull/2292/commits/d1050e806c759f3bec305267b7e9b1bc58a2f2bf)

File: src/main/kotlin/translations/sorting/TranslationTemplateConfigurable.kt [hunk 8]
```diff
@@ -61,6 +75,33 @@ class TranslationTemplateConfigurable(private val project: Project) : Configurab
...
+
+            row {
+                textField().bindText(translationSettings::convertToTranslationTemplate)
+                    .enabledIf(allowConvertToTranslationTemplate)
+            }
+
+            onApply {
+                for (project in ProjectManager.getInstance().openProjects) {
+                    ProjectView.getInstance(project).refresh()
+                }
```
Comment (RedNesto) [inline src/main/kotlin/translations/sorting/TranslationTemplateConfigurable.kt]: Why is this needed?

File: src/main/kotlin/translations/TranslationFiles.kt [hunk 9]
```diff
@@ -110,12 +111,46 @@ object TranslationFiles {
...
+            ?: throw IllegalArgumentException("Cannot add translation for element outside of module")
+        var jsonVersion = true
+        if (!TranslationSettings.getInstance(context.project).isForceJsonTranslationFile) {
+            val version =
+                context.mcVersion ?: throw IllegalArgumentException("Cannot determine MC version for element $context")
+            jsonVersion = version > MC_1_12_2
+        }
+
+        if (!jsonVersion) {
+            return null // todo: lang file support
```
Comment (RedNesto) [inline src/main/kotlin/translations/TranslationFiles.kt]: Honestly I think you can ignore lang files, the support for that format is basically a dangling remnant of now unsupported versions.

Comment (Moulberry) [reply src/main/kotlin/translations/TranslationFiles.kt]: [a6112d6](https://github.com/minecraft-dev/MinecraftDev/pull/2292/commits/a6112d6d6299f333762e64609930b7c71463b2c3)

File: src/main/kotlin/translations/sorting/TranslationTemplateConfigurable.kt (hunk 8 above)
Comment (Moulberry) [reply src/main/kotlin/translations/sorting/TranslationTemplateConfigurable.kt]: [9c4a0c0](https://github.com/minecraft-dev/MinecraftDev/pull/2292/commits/9c4a0c09a8260f79ed2a4c74ed5ff73ad0fc9d24)

File: src/main/kotlin/translations/intentions/ConvertToTranslationIntention.kt [hunk 10]
```diff
@@ -73,12 +77,19 @@ class ConvertToTranslationIntention : PsiElementBaseIntentionAction() {
...
                 if (replaceLiteral) {
+                    val translationSettings = TranslationSettings.getInstance(project)
                     val psi = PsiDocumentManager.getInstance(project).getPsiFile(editor.document) ?: return
                     psi.runWriteAction {
                         val expression = JavaPsiFacade.getElementFactory(project).createExpressionFromText(
-                            "net.minecraft.client.resources.I18n.format(\"$key\")",
+                            if (translationSettings.isUseCustomConvertToTranslationTemplate) {
+                                translationSettings.convertToTranslationTemplate.replace("\$key", key)
+                            } else {
+                                "net.minecraft.client.resources.I18n.format(\"$key\")"
```
Comment (Earthcomputer) [inline src/main/kotlin/translations/intentions/ConvertToTranslationIntention.kt:91]: Wasn't the point of adding a checkbox for a custom template to use the mappings manager?

**Review by Earthcomputer — CHANGES_REQUESTED** [review]
Looks good other than this one thing, and also formatting needs to be fixed with ktlint

File: src/main/kotlin/translations/intentions/ConvertToTranslationIntention.kt (hunk 10 above)
Comment (Moulberry) [reply src/main/kotlin/translations/intentions/ConvertToTranslationIntention.kt:91]: My understanding from your previous message is that the mappings API isn't ready yet.
I added the checkbox so that when it's ready it can be integrated without requiring users to tick another box when they update.

**Comment by RedNesto** [comment]
Thank you for your PR.

I rebased and merged it in 336924bed50a7b31ab01929da3477c5742cf9b02 with a change addressing Earth's comment.
